### ISSUE TITLE
Version 4.0.0-alpha - control of FS20 devices after restarting the laptop only works after unplugging and reconnecting the FS20

### ISSUE THREAD

**Issue by stefaniehoechtl** [asterics/AsTeRICS-Ergo#108]
AsTeRICS Ergo works perfectly on Peter’s PC (Windows 10 Pro, using Microsoft Edge) and my old laptop (Microsoft Windows 7 Home Premium, using Google Chrome).
On my laptop (Acer Switch SA5-271; Windows 10 Home, using Mozilla Firefox) setting up the control worked perfectly. However, I cannot control the FS20 devices after restarting the computer. I always need to start AsTeRICS Ergo, unplug and reconnect the FS20 (in exactly this order), and then the control of the FS20 devices works again.

**--> control of FS20 devices after restarting the laptop (Acer Switch SA5-271; Windows 10 Home) only works after unplugging and reconnecting the FS20**

**Comment by klues** [comment]
@stefaniehoechtl questions:
* in the state, where FS20 is not working - is the red light on FS20 pc sender shining or dark?
* did you do the FS20 patch on first installation?

**Comment by stefaniehoechtl** [comment]
There were no problems on first installation
The light on the FS20 PC sender was not shining
I reinstalled the FS20 today – now the control of FS20 devices work without any problems after restarting the computer – I think that is strange.

I am gonna try the latest release on another laptop today, where I had the same problem.

**Comment by stefaniehoechtl** [comment]
I still have the problem on the other laptop (Windows 10 Home, Medion Akoya E7422 MD99920, Avira, Firefox). This time I tested with AsTeRICS Ergo v4.0.0-alpha3.
After restarting the computer (or after standby-mode), the light in the FS20 pc sender is not shining. After unplugging the FS20 and reconnecting it to the computer, the light is shining again and the control of the FS20 devices works.

**Comment by klues** [comment]
> There were no problems on first installation

So you did not do the registry patch?
I sometimes had the case, that the red light keeps shining, if AsTeRICS-Ergo is started (so you think that you do not need the patch), but if AsTeRICS-Ergo is not started, the light turns off after a while. So that could be the case also at your problem:
1) on first installation the red light keeps shining -> you do not apply the patch
2) after restarting PC AsTeRICS-Ergo is not running and the light goes off

I dont know how to fix this problem - we could change the assistant so that the patch is started in any case...

**Comment by stefaniehoechtl** [comment]
In both cases I did not need to do the registry patch when installing the FS20 for the first time, because the light was still shining after 10 sec.
So yes that could be the problem.

**Comment by klues** [comment]
I think the best solution would be:
* run the patch everytime on installation wizard, do not ask the user if light is shining or not. this would be a more user-friendly way, the user would not have to wait and check.
* The patch-method detects if the patch is actually needed, so the user would only get the UAC-message if it is really needed.

an alternative would be:
maybe it also helps if we change the code for FS20 in that way, that every connection to FS20 is closed, if a command to FS20 fails. So AsTeRICS-Ergo does not has a pointer to the device anymore and the operating system maybe will send it to power-save mode - so that the installation wizard is correct and the light is not shining anymore...

but I would prefer the first suggestion, because at the second we still not know if it would work.
@deinhofer what do you think?

**Comment by stefaniehoechtl** [comment]
I just tried something:
Installation FS20 - the light was still shining after 10 sec, but I selected "Lamp is not lit" - "Fix problem". The test afterwards failed, so I had to install the FS20 another time, but now I can control all devices after restarting the computer.
Maybe that helps you a little...

**Comment by klues** [comment]
yes, its exactly what i thought. I'll change the assistant so that the patch is activated every time. i already spoke with @deinhofer and we think thats the best solution.
I'll upload a new release after changing that - so until then you do not have to test further FS20-things ;)

**Comment by deinhofer** [comment]
**go ahead with the solution: always try to patch the system if the registry key is not found!**

an idea for later:
execute the patch during postinstall similar to the firewall patch. But we would have to test if this also works if no FS20 device is plugged in.

**Comment by klues** [comment]
changed fs20 assistant with https://github.com/asterics/AsTeRICS-Ergo/pull/116

@deinhofer could you please build and publish a new release with AsTeRICS-Ergo master and AsTeRICS tag `asterics-ergo/v4.0.0-alpha4`?

**Comment by klues** [comment]
@stefaniehoechtl you can now try if the FS20-error is resolved with https://github.com/asterics/AsTeRICS-Ergo/releases/tag/v4.0.0-alpha4

**Comment by stefaniehoechtl** [comment]
I tested AsTeRICS Ergo v4.0.0-alpha5 on 5 laptops, the control of the FS20 devices always worked after restarting the laptop.

**Comment by ChrisVeigl** [comment]
is the suggested procedure now to restart the laptop after installation ?
BTW: at our presentation in Dortmund, the FS20 light went out (after dozens of times where it stayed on...) no idea why. however, the problem could be fixed using the "repair" button in the AsTeRICS Ergo Gui.

**Comment by stefaniehoechtl** [comment]
No, I do not think it is necessary to restart the laptop after installation. 
During testing, I just always tried whether the control still worked after restarting the computer, because we often had the problem that the control of the FS20 devices worked at first, but after restarting the computer, it did not work anymore. But that problem seems to be fixed now.
On the other hand it does not do any harm to check whether the control still works after restart.

**Comment by klues** [comment]
> BTW: at our presentation in Dortmund, the FS20 light went out (after dozens of times where it stayed on...) no idea why. 

I also encountered this behaviour that the light stays on most of the time if AsTeRICS-Ergo is started, but not in every case. I think it is somehow related to the implementation of the FS20 code, where some reference to the device is used and therefore the operating system thinks that the device is needed and does not send it into power-save mode... but I also do not fully understand it.

thats the reason why we now changed the assistant in a way that the registry patch is applied everytime. this should fix the problem, because it permanently disables the power-save mode in every case.
